src/heap/sharedMem.ml

```diff
@@ -1071,6 +1111,298 @@ module NewAPI = struct
+      unsafe_write_addr chunk data;
+      for i = 0 to level do
+        unsafe_write_addr chunk null_addr
+      done;
+      addr
+    in
+    (size, write)
+
+  let sklist_search_level_addr sklist = addr_offset sklist 1
+
+  let sklist_succ_addr level sklist = addr_offset sklist (2 + level)
+
+  let sklist_get_search_level = get_generic_int sklist_search_level_addr
+
+  let sknode_data_addr sknode = addr_offset sknode 1
+
+  let sknode_succ_addr = sklist_succ_addr
+
+  let sknode_get_data = get_generic sknode_data_addr
+
+  let sknode_max_level sknode =
+    let hd = read_header (get_heap ()) sknode in
+    obj_size hd - 2
+
+  let sklist_is_marked x = Int64.(equal 1L (logand x 1L))
+
+  let sklist_marked x = Int64.logor x 1L
+
+  let sklist_unmark x = Int64.(to_int (logand x (lognot 1L)))
+
+  (* Search the skip list, returning true if the a node with `data` is in the
+   * set.
+   *
+   * This function will physically delete any marked nodes by updating the
+   * node's predecessor's successor pointer to the node's successor.
+   *
+   * The caller-provided `preds` and `succs` arrays are populated with the
+   * target node's predecessors and successors at each level. These arrays are
+   * used to implement `sklist_add` and `sklist_remove`. *)
+  let sklist_find cmp sklist data preds succs =
+    let rec start () =
+      let top_level = sklist_num_levels - 1 in
+      let pred = sklist_succ_addr top_level sklist in
+      let curr = sklist_unmark (load_acquire pred) in
+      search top_level pred curr
+    and search level pred curr =
+      if curr == null_addr then
+        (* We are at the end of the list at this level. *)
+        down level pred curr
+      else
+        let succ_addr = sknode_succ_addr level curr in
+        let succ64 = load_acquire succ_addr in
+        let succ = sklist_unmark succ64 in
+        if sklist_is_marked succ64 then
+          (* `curr` is logically deleted. Try to physically remove `curr` from
+           * this level by updating `pred` to point directly to `succ`. *)
+          if compare_exchange_strong pred (i64 curr) (i64 succ) then
+            (* We successfully removed `curr`; `pred` now points to `succ`.
+             * Continue searching along this level. *)
+            search level pred succ
+          else
+            (* We failed to remove `curr`. Another thread has either logically
+             * deleted `pred`, inserted a new successor of `pred`, or inserted
+             * a new successor of `curr`. *)
+            start ()
+        else if cmp (sknode_get_data curr) data < 0 then
+          (* Continue searching across a level until we find a node with a value
+           * that is greater than or equal to the target value. *)
+          search level succ_addr succ
+        else
+          down level pred curr
+    and down level pred curr =
+      (* Record predecessor and successor of the target at this level. *)
+      preds.(level) <- pred;
+      succs.(level) <- curr;
```

Comment:
Note there are several in-efficiencies if this is in a hot loop.

- unneededs bound checking
- hidden allocation for the mutual recursive function
- `cmp` passed as a function will trigger `caml_apply`
- `compare_file_by_name` seems to trigger allocation as mentioned above

Reply:
The loop is not particularly hot. It's a log(n) search with low n. We might iterate more if the list is under contention, but contention will be rare in practice.

The bounds checking I think is well worth the cost in this context. In general I'll take bounds checking unless profiling show that it's a meaningful bottleneck...

Regarding the comparison, we can optimize for sure, assuming it makes a difference in practice. I suspect the cost of caml_apply and allocations is likely overwhelmed by the actual string comparison. We could assign each file a unique integer ID and use that instead.